contributors/design-proposals/node/limit-node-object-self-control.md

```diff
@@ -0,0 +1,128 @@
+management difficult. Instead of configuring labels and taints in a centralized
+place, we must configure `N` kubelet command lines. More significantly, the
+approach greatly compromises security. Below are two straightforward escalations
+on an initially compromised node that exhibit the attack vector.
+
+### Capturing Dedicated Workloads
+
+Suppose company `foo` needs to run an application that deals with PII on
+dedicated nodes to comply with government regulation. A common mechanism for
+implementing dedicated nodes in Kubernetes today is to set a label or taint
+(e.g. `foo/dedicated=customer-info-app`) on the node and to select these
+dedicated nodes in the workload controller running `customer-info-app`.
+
+Since the nodes self reports labels upon registration, an intruder can easily
+register a compromised node with label `foo/dedicated=customer-info-app`. The
+scheduler will then bind `customer-info-app` to the compromised node potentially
+giving the intruder easy access to the PII.
+
+This attack also extends to secrets. Suppose company `foo` runs their outward
+facing nginx on dedicated nodes to reduce exposure to the company's publicly
+trusted server certificates. They use the secret mechanism to distribute the
+serving certificate key. An intruder captures the dedicated nginx workload in
+the same way and can now use the node certificate to read the company's serving
+certificate key.
+
+## Proposed Solution
+
+In many environments, we can improve the situation by centralizing reporting of
+sensitive node attributes to a more trusted source and disallowing reporting of
+these attributes from the kubelet.
+
+### Label And Taint Restriction
+
+An operator will configure a whitelist of taints and labels that nodes are
+allowed to set on themselves. This list should include the taints and labels
+that the kubelet is already setting on itself.
+
+Well known taint keys:
+```
+node.cloudprovider.kubernetes.io/uninitialized
+```
+
+Well known label keys:
+
+```
+kubernetes.io/hostname
+failure-domain.beta.kubernetes.io/zone
+failure-domain.beta.kubernetes.io/region
+beta.kubernetes.io/instance-type
+beta.kubernetes.io/os
+beta.kubernetes.io/arch
+```
+
+As well as any taints and labels that the operator is setting using:
+
+```
+  --register-with-taints
+  --node-labels
+```
+
+This whitelist is passed as a command line flag to the apiserver.
+NodeRestriction admission control will then prevent setting and modification by
+nodes of all taints and labels with keys not in the whitelist.
+
+
+### NodeRestriction Config
+
+A new configuration API group will be created for the NodeRestriction admission
+controller with the name `noderestriction.admission.k8s.io`. It will contain one
+config object:
+
+```golang
+type Configuration struct {
+  // RestrictLabels specifies whether the admission controller should restrict
+  // node self set labels to the whitelist.
+  RestrictLabels bool
```

Comment:
What's the tradeoff between a bool and allowing `*` in the whitelist?

Reply:
Also, any way for us to make this not default to "allow all"?

Reply:
@ericchiang how do we start restricting without breaking compatibility with the API?

Reply:
@mikedanese by specifying the config you opt into the behavior of the `NodeRestriction` component? Tune it to get the old, less secure behavior back?

Reply:
I guess I'm confused how this config is being passed to kube. Is it not a flag on the API server or something?

Reply:
Ya, I'm fine with that approach. The config is passed in via the --admission-control-config-file flag on the apiserver:

https://kubernetes.io/docs/admin/admission-controllers/#configuration-file-format